### DISCUSSION TITLE
Enemy health and shield level visible when targetting

### DISCUSSION THREAD

**Discussion by mefistotelis** [swfans/syndwarsfx#172] [General]
I just fixed a bug and now additional bars are visible on targets.
The original game always tried to draw them off-screen...

Do you think these should stay? Maybe be closer? or on opposite sides of the target? change their style? or maybe only leave one?

Let me know what you think.

Related commit:
[swars: Fixed invalid coords computation when drawing target](https://github.com/swfans/swars/commit/d12e12c6c818f42ad44cb0e7e43e468c4a8a80bd)

**Comment by Moburma** [comment]
Ahhh that's a great find. Took me a while to notice what it actually is - it's extra bars for health and shield visible in the little blue targeting recticle thing when mousing over enemies. I actually always though it was a bit weird that this has an "unbalanced" shape normally, now I know why.
Pretty cool little feature, although it's hard to see, especially as the mouse movement update is slower than the original. Maybe try making the health different colours, like? 
It's good though, and I think to be honest difficult enough to use that it's not "cheating". I actually thought it'd work better in direct control mode, but it uses a different recticle and doesn't show up then, it's a mouse only feature.

**Reply by mefistotelis** [reply]
> it's hard to see, especially as the mouse movement update is slower than the original

It was likely designed for 320x200. Maybe the code was originally working, but got damaged when switching the rounded target marks into rectangular ones?

EDIT:
For the mouse, I did some changes recently, and it might be now possible. Not to make the reaction instant, but at least to make several updates per frame.

**Comment by geist22** [comment]
Just noticed these new bars, looks kind of weird so I thought it was a bug at first. Also, they are moving around slightly when the target is moving, like they're aligned to the target's body rather than the reticle.
I too find them hard to see, at least in 640x480. In 320x200 they are twice as wide compared to the reticle's borders so they're much easier to see and look alright there, though the screen being stretched might help it a bit too.

More quality of life features regarding the user interface are always nice, but the original game doesn't give you that kind of information, so it may be kind of cheaty.

In my opinion this could be a case for a config entry, rules.ini or CONFIG.INI? rules.ini feels like it's related to game mechanics, so UI stuff would belong to CONFIG.INI? Though zoom levels are in rules.ini...

Oh, and the bars really stick out on vehicles, on top of one being yellow. Do vehicles even need these? They already have their own bar, so it feels redundant here. I think vehicles should have them removed.

Really looking forward for a cursor that updates at host refresh rates (or at least 60Hz), this has always been my Nr. 1 most wanted feature (or rather, fix) for the port.

**Comment by mefistotelis** [comment]
Changes:
* Width now scales to resolution
* Not shown for vehicles

Making it an addition enabled in rules file might also be a good idea.

How does it look now?

**Comment by geist22** [comment]
Hmm... it's hard to say how it's supposed to look. The low-res one is still easier to see because it's not touching the border, while the high-res one is right next to it and also clipping through the corner at the bottom, which probably shouldn't happen.

![reticle](https://github.com/swfans/swars/assets/107389189/4c1a7807-f3c9-4488-9721-ba6e2f347c02)

Looking at this I think moving the right bar in the low-res version 1 pixel to the left, so they each are 1 pixel to the side of the reticle borders might produce the best look I can think of right now. And then match that with the high-res one.
I don't know, it kind of feels like it's designed to have the bars in the same lines as the borders of the reticle, like the left bar in the high-res version now, maybe this is the "correct" position after all, though this makes it harder to see...

The bars moving around is very bothersome though. Try looking at a stationary target and rotate the camera to see it happening.

-edit-
The longer I think about this, the more I think that the high-res bars are actually in a good position right now.

**Comment by mefistotelis** [comment]
Pre-alpha version has these bars. So they were clearly designed to blend into the target marker there.

Here, the red reveals depleted energy bar. For which line is the health bar - not sure, but it's probably there.

![obraz](https://github.com/swfans/swars/assets/11463396/2b8a1f13-5ae8-42ce-8fe4-da3d4b83622c)

**Comment by geist22** [comment]
Bunch of shopped tests:

Idea stated above, also moved bars up a bit to avoid clipping and match spacing from reticle border:
![01](https://github.com/swfans/swars/assets/107389189/d940b553-b3bb-491c-97ae-f4a7604a873d)

Moved both bars half a bar closer in both directions:
![02](https://github.com/swfans/swars/assets/107389189/3ec55478-7a43-4e62-adca-a6e21efc96dd)

Current but with right bar moved closer:
![03](https://github.com/swfans/swars/assets/107389189/de33332a-88bc-4b22-a808-b0f688e7100b)

**Comment by mefistotelis** [comment]
I wonder if we can bring the whole pre-alpha frame back..

It would require:
1. Having the sprites `mspr-0.dat`
2. Having sprites mentioned as Elements in `mele-0.ani`
3. Having Elements binded into animation frames in `nfra-0.ani`

Currently, I don't think `nfra-0.ani` contains the required animation frames.

**Comment by mefistotelis** [comment]
Looked more into it.
So we need the selected sprites to make the old frame:
![obraz](https://github.com/swfans/swars/assets/11463396/89641c33-d6b0-4bce-a77b-0923c9a5879a)

In pre-alpha, these are id=507 .. 533.
They do exist in final. There, they're id=969 .. 995.

Here are dumpred pre-alpha Frames and Elements which use these sprites:

```
F1045  {2744,48,78, 0, 0x01,1046},
E02744    {3042,-20,-48, 0x30,2745},
E02745    {3096,-12,-58, 0x30,2746},
E02746    {3144,-24,-40, 0x30,2747},
E02747    {3072, 18,-38, 0x30,2748},
E02748    {3042, 12, 12, 0x30,2750},
E02750    {3036, 18,-30, 0x34,   0},
F1046  {2700,48,78, 0, 0x00,1047},
E02700    {3048,-20,-48, 0x30,2701},
E02701    {3048, 12, 12, 0x30,2718},
E02718    {3102,-12,-58, 0x30,2727},
E02727    {3150,-24,-40, 0x30,2736},
E02736    {3078, 18,-38, 0x30,2751},
E02751    {3036, 18,-30, 0x34,   0},
F1047  {2702,48,78, 0, 0x00,1048},
E02702    {3054,-20,-48, 0x30,2703},
E02703    {3054, 12, 12, 0x30,2719},
E02719    {3108,-12,-58, 0x30,2728},
E02728    {3156,-24,-40, 0x30,2737},
E02737    {3084, 18,-38, 0x30,2752},
E02752    {3036, 18,-30, 0x34,   0},
F1048  {2704,48,78, 0, 0x00,1049},
E02704    {3060,-20,-48, 0x30,2705},
E02705    {3060, 12, 12, 0x30,2720},
E02720    {3114,-12,-58, 0x30,2729},
E02729    {3162,-24,-40, 0x30,2738},
E02738    {3090, 18,-38, 0x30,2753},
E02753    {3036, 18,-30, 0x34,   0},
F1049  {2706,48,78, 0, 0x00,1050},
E02706    {3066,-20,-48, 0x30,2707},
E02707    {3066, 12, 12, 0x30,2721},
E02721    {3120,-12,-58, 0x30,2730},
E02730    {3168,-24,-40, 0x30,2739},
E02739    {3072, 18,-38, 0x30,2754},
E02754    {3036, 18,-30, 0x34,   0},
F1050  {2708,48,78, 0, 0x00,1051},
E02708    {3042,-20,-48, 0x30,2709},
E02709    {3042, 12, 12, 0x30,2722},
E02722    {3126,-12,-58, 0x30,2731},
E02731    {3174,-24,-40, 0x30,2740},
E02740    {3072, 18,-38, 0x30,2755},
E02755    {3036, 18,-30, 0x34,   0},
F1051  {2710,48,78, 0, 0x00,1052},
E02710    {3048,-20,-48, 0x30,2711},
E02711    {3048, 12, 12, 0x30,2723},
E02723    {3132,-12,-58, 0x30,2732},
E02732    {3180,-24,-40, 0x30,2741},
E02741    {3072, 18,-38, 0x30,2756},
E02756    {3036, 18,-30, 0x34,   0},
F1052  {2712,48,78, 0, 0x00,1053},
E02712    {3054,-20,-48, 0x30,2713},
E02713    {3054, 12, 12, 0x30,2724},
E02724    {3138,-12,-58, 0x30,2733},
E02733    {3186,-24,-40, 0x30,2742},
E02742    {3072, 18,-38, 0x30,2757},
E02757    {3036, 18,-30, 0x34,   0},
F1053  {2714,48,78, 0, 0x00,1054},
E02714    {3060,-20,-48, 0x30,2715},
E02715    {3060, 12, 12, 0x30,2725},
E02725    {3096,-12,-58, 0x30,2734},
E02734    {3192,-24,-40, 0x30,2743},
E02743    {3072, 18,-38, 0x30,2758},
E02758    {3036, 18,-30, 0x34,   0},
F1054  {2716,48,78, 0, 0x00,1045},
E02716    {3066,-20,-48, 0x30,2717},
E02717    {3066, 12, 12, 0x30,2726},
E02726    {3096,-12,-58, 0x30,2735},
E02735    {3198,-24,-40, 0x30,2749},
E02749    {3072, 18,-38, 0x30,2759},
E02759    {3036, 18,-30, 0x34,   0},
```

In this dump, every Element refers a sprite id. But to make the thing more complicated, the id is multiplied by 6. Ie. the first element in the list above has the value 3042, so it uses sprite id 3042 / 6 = 507.
Last number in each line is always the next item - for frames next Frame, for elements next Element.

Now need to check whether the Elements exist in final.

**Comment by mefistotelis** [comment]
I made the old frame work. But it only matches one bar - in pre-alpha there was only one after all.

![obraz](https://github.com/swfans/swars/assets/11463396/c65517b6-9926-4b5b-853c-3771bef5bc4f)

Also the `mele` and `nfra` files need new entries for this to work..
So I'm not sure whether this should be enabled by the same option as the bars. Maybe the old frame should be a separate new option.

**Reply by mefistotelis** [reply]
@Moburma can you add gfx package downloading to the installer, in the same way levels are downloaded?

I added updated ANI files to the `swars-gfx` repo.

**Reply by Moburma** [reply]
Yep, will do
